Feature external api integration

pachecowolox opened the pull request:
## Summary
- Add feing client
## Trello Card
https://trello.com/c/Fy4fLwEr/22-integraci%C3%B3n-api-externa

pachecowolox committed:
change version spring boot to use open-feing

pachecowolox committed:
Merge branch 'feature-lombook-implementation' into feature-external-api-integration

# Conflicts:
#	src/test/java/wolox/training/controller/BookControllerTest.java
#	src/test/java/wolox/training/controller/UserControllerTest.java

pachecowolox committed:
add feing client ,
add BookInfoDto

pachecowolox committed:
add mockBean feing in BookControllerTest

pachecowolox committed:
add feing client,refactor findByBookIsb external service, clean imports

pachecowolox committed:
Merge branch 'development' into feature-external-api-integration

pachecowolox committed:
add exception data not found

marcegl91 commented on src/main/java/wolox/training/client/delegate/OpenLibraryDelegate.java:
```diff
@@ -0,0 +1,66 @@
+		book.setImage("img");
+		book.setGenre("M");
+		book.setTitle(bookInfoDto.getTitle());
+		if (bookRepository.findByIsbn(book.getIsbn()).isPresent()) {
```
Why do you need this verification? If I am not mistaken, when you called the method  `OpenLibraryDelegate.findBookByIsbn` you already checked that the book was not found in the database.

marcegl91 commented on src/main/resources/application.properties:
```diff
@@ -5,4 +5,6 @@ spring.jpa.properties.hibernate.dialect = org.hibernate.dialect.PostgreSQLDialec
 server.port = 8081
 spring.devtools.restart.log-condition-evaluation-delta=false
-
+feign.placeholder.name= BOOKS
```
Please use more descriptive names than `placeholder`.

marcegl91 commented on src/main/java/wolox/training/controllers/BookController.java:
```diff
@@ -110,4 +130,24 @@ public Book updateBook(@RequestBody Book book, @PathVariable Long id) {
+	public ResponseEntity<Book> findBookByIsbn(@PathVariable String isbn) {
+
+		Optional<Book> book = bookRepository.findByIsbn(isbn);
+		if (book.isPresent()) {
```
You can improve the readability of this code:

```
bookRepository.findByIsbn(isbn)
.map(book -> new ResponseEntity<>(book,HttpStatus.OK))
.orElseGet(new ResponseEntity<>(openLibraryDelegate.findBookByIsbn(isbn).get(),HttpStatus.CREATED));
```

Try it and let us know how it worked in both cases (when the book is found in the repository and when it is not found and has to retrieve it to the external API)

Also, is there any need for openLibraryDelegate.findBookByIsbn to return an Optional? I don't think you are using anything related to Optional in that method, you can return directly the Book.

pachecowolox committed:
change application properties publisher to library,
change return findBookByIsbn to lambda expression,
change return Optional book to Object Book in the method saveBookOpenLibrary

marcegl91 commented on src/main/java/wolox/training/controllers/BookController.java:
```diff
@@ -110,4 +130,17 @@ public Book updateBook(@RequestBody Book book, @PathVariable Long id) {
+
+	@GetMapping("/isbn/{isbn}")
+	public ResponseEntity<Book> findBookByIsbn(@PathVariable String isbn) {
+		return bookRepository.findByIsbn(isbn).map(book -> new ResponseEntity<>(book, HttpStatus.OK))
```
You can improve the readibility of this code by doing something like this:

```
return bookRepository
           .findByIsbn(isbn)
           .map(book -> new ResponseEntity<>(book, HttpStatus.OK))
	   .orElseGet(() -> new ResponseEntity<>(openLibraryDelegate.findBookByIsbn(isbn), HttpStatus.CREATED));
```

marcegl91 reviewed (approved):
Great job!
